crates/next-core/src/page_loader.rs

```diff
@@ -0,0 +1,153 @@
+};
+
+use crate::embed_js::next_js_file;
+
+#[turbo_tasks::function]
+pub async fn create_page_loader(
+    server_root: FileSystemPathVc,
+    client_context: AssetContextVc,
+    client_chunking_context: ChunkingContextVc,
+    entry_asset: AssetVc,
+    pathname: StringVc,
+) -> Result<ContentSourceVc> {
+    let asset = PageLoaderAsset {
+        server_root,
+        client_context,
+        client_chunking_context,
+        entry_asset,
+        pathname,
+    }
+    .cell();
+
+    Ok(AssetGraphContentSourceVc::new_lazy(server_root, asset.into()).into())
+}
+
+#[turbo_tasks::value(shared)]
+pub struct PageLoaderAsset {
+    pub server_root: FileSystemPathVc,
+    pub client_context: AssetContextVc,
+    pub client_chunking_context: ChunkingContextVc,
+    pub entry_asset: AssetVc,
+    pub pathname: StringVc,
+}
+
+#[turbo_tasks::value_impl]
+impl PageLoaderAssetVc {
+    #[turbo_tasks::function]
+    async fn get_loader_entry_asset(self) -> Result<AssetVc> {
+        let this = &*self.await?;
+
+        let mut result = RopeBuilder::default();
+        writeln!(
+            result,
+            "const PAGE_PATH = {};\n",
+            stringify_str(&format!("/{}", &*this.pathname.await?))
+        )?;
+
+        let base_code = next_js_file("entry/page-loader.ts");
+        if let FileContent::Content(base_file) = &*base_code.await? {
+            result += base_file.content()
+        } else {
+            bail!("required file `entry/page-loader.ts` not found");
+        }
+
+        let file = File::from(result.build());
+
+        Ok(VirtualAssetVc::new(this.entry_asset.path().join("page-loader.ts"), file.into()).into())
+    }
+
+    #[turbo_tasks::function]
+    async fn get_page_chunks(self) -> Result<ChunksVc> {
+        let this = &*self.await?;
+
+        let loader_entry_asset = self.get_loader_entry_asset();
+
+        let asset = EcmascriptModuleAssetVc::new(
+            loader_entry_asset.into(),
+            this.client_context,
+            turbo_tasks::Value::new(EcmascriptModuleAssetType::Typescript),
+            EcmascriptInputTransformsVc::cell(vec![EcmascriptInputTransform::TypeScript]),
+            this.client_context.environment(),
+        );
+
+        let chunk_group =
+            ChunkGroupVc::from_chunk(asset.as_evaluated_chunk(this.client_chunking_context, None));
+
+        Ok(chunk_group.chunks())
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl Asset for PageLoaderAsset {
+    #[turbo_tasks::function]
+    async fn path(&self) -> Result<FileSystemPathVc> {
+        Ok(self
+            .server_root
+            .join("_next/static/chunks/pages")
+            .join(&get_asset_path_from_route(&*self.pathname.await?, ".js")))
+    }
+
+    #[turbo_tasks::function]
+    async fn content(self_vc: PageLoaderAssetVc) -> Result<AssetContentVc> {
+        let this = &*self_vc.await?;
+
+        let chunks = self_vc.get_page_chunks();
+
+        let mut data = Vec::new();
```

Comment:
You could allocate a vector of the right capacity right away with `chunks.await?.len()`.

Reply:
with formatting that turns into
```rs
        let mut data = chunks
            .await?
            .iter()
            .map(|&chunk| async move {
                let path = chunk.path().await?;
                Ok(Value::String(format!("/{}", path.path)))
            })
            .try_join()
            .await?;
```
which imo is way harder to read

Reply:
Added vector pre allocation for now